### DISCUSSION TITLE
Pytest-bdd integration with selenium base

### DISCUSSION THREAD

**Discussion by taruntechno** [seleniumbase/SeleniumBase#1200] [Q&A] [answered]
Hello Team,
I need your help in knowing how to integrate pytest-bdd level with selenium base.
My feature file is 
`@speedtest
Feature: Test Internet Speed
  As a user I want to test the internet speed

Background:
  Given Bell Speed test page shows up

Scenario: Speed Test
  When User clicks on check speed button
  Then Speed details are saved
`

Step Defs is 
`from seleniumbase import BaseCase
import time
import pytest
from pytest_bdd import scenarios, given, when, then, parsers

scenarios('../features/speedtest.feature')

class SpeedTest(BaseCase):
  @given('Bell Speed test page shows up')
  def open_bell_speed_test(self):
    self.open("https://support.bell.ca/internet/internet-speed-test")

  @when('User clicks on check speed button')
  def start_speed_test(self):
    self.click('button[id="skst-enh-init-test-btn"]')

  @then('Speed details are saved')
  def test_bell_internet(self):
    time.sleep(120)
    self.save_screenshot("speedtest.png", folder="./downloaded_files")
    self.assert_downloaded_file("speedtest.png")
`

When I run pytest it opens the browser but does nothing . If I run the code without pytest-bdd using following it works
`class SpeedTest(BaseCase):
  def open_bell_speed_test(self):
    self.open("https://support.bell.ca/internet/internet-speed-test")
    self.click('button[id="skst-enh-init-test-btn"]')
    self.save_screenshot("speedtest.png", folder="./downloaded_files")
    self.assert_downloaded_file("speedtest.png")`

Folder structure
<img width="228" alt="CleanShot 2022-02-04 at 18 57 04@2x" src="https://user-images.githubusercontent.com/99055398/152618576-16797198-46b0-483a-b165-d097f6b6ef7a.png">

**Comment by mdmintz** [comment, marked as answer]
Hi @taruntechno, when using https://github.com/pytest-dev/pytest-bdd with SeleniumBase, you may need to use the SeleniumBase ``sb`` fixture format for compatibility. (Numbers 3 and 4 of [syntax_formats](https://github.com/seleniumbase/SeleniumBase/blob/master/help_docs/syntax_formats.md))

Here's what the ``sb`` fixture format looks like, shown below: (*pytest-bdd can be added onto that separately*)

```python
# "sb" pytest fixture test in a method with no class
def test_sb_fixture_with_no_class(sb):
    sb.open("https://google.com/ncr")
    sb.type('input[title="Search"]', "SeleniumBase GitHub\n")
    sb.click('a[href*="github.com/seleniumbase/SeleniumBase"]')
    sb.click('a[title="seleniumbase"]')

# "sb" pytest fixture test in a method inside a class
class Test_SB_Fixture:
    def test_sb_fixture_inside_class(self, sb):
        sb.open("https://google.com/ncr")
        sb.type('input[title="Search"]', "SeleniumBase GitHub\n")
        sb.click('a[href*="github.com/seleniumbase/SeleniumBase"]')
        sb.click('a[title="examples"]')
```

**Reply by taruntechno** [reply]
Wow you are so good so fast reply. I was able to figure it out from your docs and it worked like a charm. Another thing I was wondering while exploring is when I run pytest, it opens a browser instance and then closes it immideatly and opens another one and then does the test. Is there a way to avoid this.

**Reply by mdmintz** [reply]
@taruntechno That could mean that the version of chromedriver installed does not match the version of Chrome installed. You can either install a matching version with ``sbase install chromedriver NUM`` eg: ``sbase install chromedriver 98``, or you can run the driver upgrade script: https://github.com/seleniumbase/SeleniumBase/blob/master/examples/upgrade_chromedriver.py

**Reply by taruntechno** [reply]
Even after matching the version both are on 98 same behaviour. I am on macos btw. @mdmintz

**Reply by mdmintz** [reply]
@taruntechno Did you make any local changes to [seleniumbase/core/browser_launcher.py](https://github.com/seleniumbase/SeleniumBase/blob/master/seleniumbase/core/browser_launcher.py)? If there's an issue, you'd be able to track it down there via breakpoints. Likely a Chrome option wasn't valid, so it reverted to no special options (and then you would see a message such as "this browser is being controlled by automated test software", etc. Did that message appear when your test finally ran on Chrome in the second browser?

**Reply by taruntechno** [reply]
@mdmintz No local changes were made to this . I am not seeing this message "this browser is being controlled by automated test software" I am attaching the screenrecording of this

https://user-images.githubusercontent.com/99055398/152621378-f933486e-5bbb-431d-8129-40b1e26c2e01.mov

.

**Reply by mdmintz** [reply]
@taruntechno  It appears you ran a very short test that first time.
Run with ``pytest -v`` to see if you can see the name.

**Reply by taruntechno** [reply]
@mdmintz it seems to be running two tests and i dont think its respecting bdd
As first test method it ran is part of "then" statement which should run in the end.

collected 2 items

tests/step_defs/test_speed.py::test_bell_internet PASSED [ 50%]
tests/step_defs/test_speed.py::test_speed_test PASSED [100%]

**Comment by taruntechno** [comment]
@mdmintz  it seems to be running two tests and i dont think its respecting bdd 
As first test method it ran is part of "**then**" statement which should run in the end.

collected 2 items                                                                              

tests/step_defs/test_speed.py::test_bell_internet PASSED                                 [ 50%]
tests/step_defs/test_speed.py::test_speed_test PASSED                                    [100%]

**Reply by taruntechno** [reply]
@mdmintz  Was able to fix it finally as i had my step_Defs methods have test in there name which was confusing pytest. After removing test from the name it was able to run properly. Thanks a lot for all your help really appreciated.

**Reply by mdmintz** [reply]
Awesome!
